src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt

```diff
@@ -31,28 +32,46 @@ class RsUnresolvedReferenceInspection : RsLocalInspectionTool() {
     override fun buildVisitor(holder: RsProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor =
         object : RsVisitor() {
             override fun visitPath(path: RsPath) {
-                if (path.parent is RsPath) return
-                val (basePath, candidates) = AutoImportFix.findApplicableContext(holder.project, path) ?: return
+                // here we check if this path is resolved
+                // if it is we return
+                // if not than check whether qualifier of this path is resolved
+                // if it is we register problem
+                // if not then we assume that unresolved parent paths of this path
+                // were or will be handled
+                // because all paths will eventually be processed by visitor
+                val (problemPath, candidates) = AutoImportFix.findApplicableContext(holder.project, path)
```

Comment:
Hm, it doesn't make sense to look for auto import candidates for each path because in path `foo::bar::baz` we can import only `foo` part, i.e. only base path. If `bar` cannot be resolved and `foo` can, we definitely should highlight `bar` as unresolved but without auto-import quick fix.
I think we should use `AutoImportFix.findApplicableContext` only when `path == path.baseBath`.

It seems we should do the following things:
* check that reference cannot be resolved
* check that its child path can be resolved (it's cheap because resolve result is already in the cache). If no, just do nothing. Otherwise, highlight it as unresolved
* if path is the same as base bath, try to find import candidates and do all related stuff as before

Probably, some code from `AutoImportFix.findApplicableContext` should be extraced to `RsUnresolvedReferenceInspection` itself.

Reply:
So, if I correctly understand, you ask me to separate checking if path is resolved and actually getting import candidates. I don't fully know intellij api and code is spread out so it's a little hard for me to fully wrap my head around this. One thing preventing me from easily splitting resolve and lookup is that lookup is done in 2 different places:
https://github.com/intellij-rust/intellij-rust/blob/9ccbb03af96b5804126b4ff555280635afbfc503/src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt#L103
and
https://github.com/intellij-rust/intellij-rust/blob/9ccbb03af96b5804126b4ff555280635afbfc503/src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt#L121
Any pointers here?

Reply:
> So, if I correctly understand, you ask me to separate checking if path is resolved and actually getting import candidates.

Not quite. You don't need to invoke `AutoImportFix.getApplicableContext()` where it doesn't make sense.

You need to check that path cannot be resolve to single item (`path.reference.resolve() == null`)
In this case, you need to check if this path has qualifier (`path.qualifier`):
* if `path.qualifier == null` (i.e. `path == path.basePath()`) you need to check if `AutoImportFix` can be applied to it (i.e. make the same actions that currently in master)
* if `path.qualifier != null`, check if it can be resolved to single item. And only of it can be resolved, you need to add error annotation about unresolved reference for `path`. Also, you don't need to add `AutoImportFix` here because it doesn't make sense

Looks like all checks described above don't require changes in `AutoImportFix`

Reply:
Wouldn't `path.reference.resolve() == null` cause false positives/negatives? Like primitives for example would resolve to null. `AutoImportFix.findApplicableContext` already handles this and couple other cases. That is why I was asking separating that method.

Reply:
Yeah, I've forgotten about primitive types. But looks like it's only one additional thing that you need to check

Reply:
There is a comment that says that some traits that would be resolved by plugin that wouldn't be resolved by the rustc. Which, if not handled, would cause false negatives

https://github.com/intellij-rust/intellij-rust/blob/9ccbb03af96b5804126b4ff555280635afbfc503/src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt#L99